src/openrct2/drawing/Snow.cpp

```diff
@@ -0,0 +1,99 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2019 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#include "../config/Config.h"
+#include "../interface/Viewport.h"
+#include "../ride/TrackDesign.h"
+#include "../scenario/Scenario.h"
+#include "../ui/UiContext.h"
+#include "../world/Climate.h"
+#include "Drawing.h"
+#include "IDrawingEngine.h"
+#include "Rain.h"
+
+using namespace OpenRCT2;
+using namespace OpenRCT2::Drawing;
+
+static void DrawLightSnow(IRainDrawer* rainDrawer, int32_t left, int32_t top, int32_t width, int32_t height);
+static void DrawHeavySnow(IRainDrawer* rainDrawer, int32_t left, int32_t top, int32_t width, int32_t height);
+
+const DrawSnowFunc DrawSnowFunctions[] = {
+    nullptr,
```

Comment:
I assume this exists for when you _don't_ want to draw snow, but wouldn't it be better to not call the DrawXSnow functions at all in these cases, instead of relying on this nullptr value?

Reply:
DrawRainFunctions also has the same nullptr value and since snow is running alongside it using IRainDrawer it just made sense to keep it.

```
const DrawRainFunc DrawRainFunctions[] = {
    nullptr,
    &DrawLightRain,
    &DrawHeavyRain,
}; 
```

Snow is also using RAIN_LEVEL_LIGHT and RAIN_LEVEL_HEAVY. 
```
const WeatherState ClimateWeatherData[9] = {
    { 10, WEATHER_EFFECT_NONE, 0, RAIN_LEVEL_NONE, SPR_WEATHER_SUN },         // Sunny
    { 5, WEATHER_EFFECT_NONE, 0, RAIN_LEVEL_NONE, SPR_WEATHER_SUN_CLOUD },    // Partially Cloudy
    { 0, WEATHER_EFFECT_NONE, 0, RAIN_LEVEL_NONE, SPR_WEATHER_CLOUD },        // Cloudy
    { -2, WEATHER_EFFECT_RAIN, 1, RAIN_LEVEL_LIGHT, SPR_WEATHER_LIGHT_RAIN }, // Rain
    { -4, WEATHER_EFFECT_RAIN, 2, RAIN_LEVEL_HEAVY, SPR_WEATHER_HEAVY_RAIN }, // Heavy Rain
    { 2, WEATHER_EFFECT_STORM, 2, RAIN_LEVEL_HEAVY, SPR_WEATHER_STORM },      // Thunderstorm
    { -10, WEATHER_EFFECT_SNOW, 1, RAIN_LEVEL_LIGHT, SPR_WEATHER_SNOW },      // Snow
    { -15, WEATHER_EFFECT_SNOW, 2, RAIN_LEVEL_HEAVY, SPR_WEATHER_SNOW },      // Heavy Snow
    { -20, WEATHER_EFFECT_BLIZZARD, 2, RAIN_LEVEL_HEAVY, SPR_WEATHER_SNOW },  // Blizzard
};
```

```
        uint32_t rainType = gClimateCurrent.RainLevel;
        if (rainType != RAIN_LEVEL_NONE && !gTrackDesignSaveMode && !(viewFlags & VIEWPORT_FLAG_HIGHLIGHT_PATH_ISSUES))
        {
            auto drawFunc = DrawSnowFunctions[rainType];

```
I guess this just saves a lot of duplication in terms of copying most of the pre-existing rain drawing code.